lib/compiler/src/beam_ssa_codegen.erl

```diff
@@ -1910,21 +1913,24 @@ collect_chain([{move,Src,Same}=M|Ms0], [{move,Same,_}|_]=Path, Others, ScrReg) -
         false ->
             collect_chain(reverse(Others, Ms0), [M|Path], [], ScrReg);
         true ->
-            %% There is a cycle, which we must break up.
+            %% There is a cycle.
             {break_up_cycle(M, Path, ScrReg),reverse(Others, Ms0)}
     end;
 collect_chain([M|Ms], Path, Others, ScrReg) ->
     collect_chain(Ms, Path, [M|Others], ScrReg);
 collect_chain([], Path, Others, _) ->
     {Path,Others}.
 
-break_up_cycle({move,Src,_}=M, Path, ScrReg) ->
-    [{move,ScrReg,Src},M|break_up_cycle1(Src, Path, ScrReg)].
+break_up_cycle({move,Src,_Dst}=M, Path, ScrReg) ->
+    break_up_cycle_1(Src, [M|Path], ScrReg, []).
 
-break_up_cycle1(Dst, [{move,Src,Dst}|Path], ScrReg) ->
-    [{move,Src,ScrReg}|Path];
-break_up_cycle1(Dst, [M|Path], LastMove) ->
-    [M|break_up_cycle1(Dst, Path, LastMove)].
+break_up_cycle_1(Dst, [{move,_Src,Dst}|Path], _SrcReg, Acc) ->
+    reverse(Acc, Path);
+break_up_cycle_1(Dst, [{move,S,D}|Path], ScrReg, Acc) ->
+    %% swap_temp will later be translated to a swap instruction
+    %% (for OTP 23 and later), or to a sequence of move instructions
+    %% (for OTP 22 and earlier).
+    break_up_cycle_1(Dst, Path, ScrReg, [{swap_temp,S,D,ScrReg}|Acc]).
```

Comment:
Is there a reason this is not done in a way similar to how the new `put_tuple` instruction is done - always emitting the new instruction and then having a pass that optionally converts it back to the old one if needed.

Reply:
Yes. Because I already have a suitable scratch register available when generating the swap instruction it was easier to it this way instead of having `beam_clean` figure out which scratch register to use.

I suppose if `beam_clean` would always use `{x,1022}` as the scratch register, I could generate `swap` instructions directly and slightly simplify `beam_ssa_codegen`, `beam_block`, and `beam_clean`. That is probably a good idea. What do you think?

Reply:
If the goal of the compatibility options is only to provide the basic compatibility without caring much about performance, I think optimising for simplicity of the compiler code is a good heuristic.

Reply:
> If the goal of the compatibility options is only to provide the basic compatibility without caring much about performance, I think optimising for simplicity of the compiler code is a good heuristic.

Yes, that is indeed the goal of the options. We need them for testing backward compatibility.

I have pushed a fixup commit.